### PR TITLE
[MAINTENANCE] Store Refactor - cloud store return types & http-errors

### PR DESCRIPTION
**Pull request by Kilo59**
Split from https://github.com/great-expectations/great_expectations/pull/5664

Changes proposed in this pull request:
- `TypedDict` return for `GeCloudStoreBackend._get()` 
   - changing return types for the other CRUD methods would require additional changes outside of `ge_cloud_store_backend.py`, this is planned but being done elsewhere.
- [adding default `timeout` values to `request` calls to prevent hanging until the connection closes](https://stackoverflow.com/a/17782541/6304433)
- always throw `StoreBackendError` on HTTP errors
- type-checking for `data_context/store/ge_cloud_store_backend.py`
- add and update Cloud related fixtures
  - new `shared_called_with_request_kwargs` fixture
  - updated `MockResponse` to support `.raise_for_status()`
  - re-use `MockResponse` instead of redefining it

## TODO
- [x] - update mocks to account for `.raise_for_status_call()`

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- ~I have made corresponding changes to the documentation~
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: tests/data_context/conftest.py [hunk 1]
```diff
@@ -633,3 +634,23 @@ def datasource_config() -> DatasourceConfig:
...
     )
+
+
+@pytest.fixture
+def bearer_test_token() -> str:
+    return "6bb5b6f5c7794892a4ca168c65c2603e"
+
+
+@pytest.fixture
+def shared_called_with_request_kwargs(bearer_test_token) -> dict:
```
Comment (Kilo59) [inline tests/data_context/conftest.py]: This is pretty verbose.
Would this make sense as just ... `called_with_request_kwargs`?

File: .gitignore [hunk 2]
```diff
@@ -64,6 +64,7 @@ nosetests.xml
 coverage.xml
 *.cover
 .hypothesis/
+src/dgtest/
```
Comment (cdkini) [inline .gitignore:67]: Why is this necessary? Are we cloning `dgtest` still?

File: great_expectations/data_context/store/ge_cloud_store_backend.py [hunk 3]
```diff
@@ -53,8 +70,8 @@ class GeCloudStoreBackend(StoreBackend, metaclass=ABCMeta):
         },
     }
 
-    RESOURCE_PLURALITY_LOOKUP_DICT: bidict = bidict(
-        **{
+    RESOURCE_PLURALITY_LOOKUP_DICT: bidict = bidict(  # type: ignore[misc]
```
Comment (cdkini) [inline great_expectations/data_context/store/ge_cloud_store_backend.py]: when we use the `# type: ignore` comment, should we add some context around why we're skipping the check? I'm good either way but we should probably establish a standard for the team.

File: great_expectations/data_context/store/ge_cloud_store_backend.py [hunk 4]
```diff
@@ -72,6 +89,8 @@ class GeCloudStoreBackend(StoreBackend, metaclass=ABCMeta):
 
     DEFAULT_BASE_URL: str = "https://app.greatexpectations.io/"
 
+    TIMEOUT: int = 20
```
Comment (cdkini) [inline great_expectations/data_context/store/ge_cloud_store_backend.py:105]: ❤️

File: great_expectations/data_context/store/ge_cloud_store_backend.py [hunk 5]
```diff
@@ -227,7 +263,7 @@ def allowed_set_kwargs(self) -> Set[str]:
             self.ge_cloud_resource_type, set()
         )
 
-    def validate_set_kwargs(self, kwargs: dict) -> Optional[bool]:
+    def validate_set_kwargs(self, kwargs: dict) -> Union[bool, None]:
```
Comment (cdkini) [inline great_expectations/data_context/store/ge_cloud_store_backend.py:279]: Should we replace all instances here? Perhaps something to discuss with the team

File: .gitignore (hunk 2 above)
Comment (Kilo59) [reply .gitignore:67]: I was running it locally in a container and it spit out this artifact.
I didn't clone it, I installed it in pip with vcs install.

Comment (cdkini) [reply .gitignore:67]: huh is that to be expected? I'm cool with it but just not what I expected.

File: great_expectations/data_context/store/ge_cloud_store_backend.py (hunk 3 above)
Comment (Kilo59) [reply great_expectations/data_context/store/ge_cloud_store_backend.py]: Yes, this particular error for `bidict` is `Keywords must be string`.
I think we actually need 2 different standards. One as we are trying to get full type coverage & another once we have full coverage.

Once we have full type coverage I think this should be the standard.
But we are sitting on over 2k typing errors now, so I think requiring this extra step now will mean it never gets done.

File: great_expectations/data_context/store/ge_cloud_store_backend.py [hunk 6]
```diff
@@ -13,9 +13,26 @@
...
+class PayloadDataField(TypedDict):
+    attributes: dict
+    id: str
+    type: str
+
+
+class ResponsePayload(TypedDict):
+    data: PayloadDataField
+    jsonapi: dict
+    links: dict
```
Comment (Kilo59) [inline great_expectations/data_context/store/ge_cloud_store_backend.py]: @wookasz `links` & `jsonapi` fields are going away, so I'll be removing them from this `TypeDict`.
https://github.com/great-expectations/great_expectations/pull/5748#discussion_r943981313
https://github.com/great-expectations/great_expectations/pull/5748#discussion_r943981511
https://github.com/great-expectations/great_expectations/pull/5748#discussion_r943981625

File: great_expectations/data_context/store/ge_cloud_store_backend.py (hunk 5 above)
Comment (Kilo59) [reply great_expectations/data_context/store/ge_cloud_store_backend.py:279]: Yes, I think this along with https://github.com/great-expectations/great_expectations/pull/5738 should be part of a custom linter/style-checker.

I will update our confluence typing page explaining these styles and why we want to use them.
Open for discussion/debate of course.

File: .gitignore (hunk 2 above)
Comment (Kilo59) [reply .gitignore:67]: I can't say for sure. But this was within a clean docker container.

**Review by cdkini — APPROVED** [review]
